design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,290 @@
+
+
+## Detailed Design
+
+### Architecture
+
+![Alt text](Prometheus_Metrics_Design.png?raw=true "Prometheus Integration Design")
+
+#### Text description
+
+1. The initializer of the consumer package calls newMetrics, a helper method 
+   that talks to Kanister’s metrics package. The result is a new metrics struct
+   that owns all the Prometheus metrics.
+
+2. In order to initialize all the required Prometheus metrics, the new_metrics 
+   method calls the InitCounterVec, InitGaugeVec, InitHistogramVec in the 
+   metrics package. It passes the metric names and the specific label names 
+   and label values as BoundedLabels to the metrics package. Once it 
+   initializes all the Prometheus metrics successfully, it returns a struct that 
+   wraps all the metrics and that the consumer package can then use. 
+
+3. The metrics package internally initializes the Prometheus metrics and 
+   registers them Prometheus. If the registration fails because the specific 
+   metric with label header already exists, the metric will simply be returned 
+   to the caller. If the registration fails due to other reasons, then the 
+   metrics package will cause a panic, signaling programmer error. 
+   In case of the CounterVec, the InitCounterVec function generates all 
+   possible permutations of label values and initializes each counter 
+   within the CounterVec with a value of 0.
+
+4. Once the collector is created in the metrics package, it will be returned 
+   to the consumer package’s newMetrics helper method.
+
+5. Once the initialization of all Prometheus metrics are complete, a new 
+   metrics struct will be returned to the consumer’s package initializer. 
+
+6. The consumer package may find it useful to implement a helper method that constructs
+   a prometheus.Labels mapping to access a specific counter from a CounterVec 
+   and perform an increment operation.
+
+
+
+### APIs
+
+#### Metrics Package
+
+```golang
+// BoundedLabel is a type that represents a label and its associated 
+// valid values
+type BoundedLabel struct {
+	LabelName   string
+	LabelValues []string
+}
+```
+
+An example of a BoundedLabel is in the scenario of ActionSet resolutions.
+Suppose we want to track these resolutions across different blueprints, 
+we would create the bounded labels in the following way:
+
+##### BoundedLabel example
+```golang
+BoundedLabel {
+  LabelName: "operation_type"
+  LabelValues: ["backup", "restore"]
+}
+
+BoundedLabel {
+  LabelName: "action_set_resolution"
+  LabelValues: ["success", "failure"]
+}
+``` 
+##### Initialization methods
+```golang
+ 
+// InitCounterVec initializes and registers the counter metrics vector. It takes a list of 
+// BoundedLabel objects - if any label value or label name is nil, then this method will panic. 
+// Based on the combinations returned by generateCombinations, it will set each counter value to 0.
+// If a nil counter is returned during registration, the method will
+// panic
```

Comment:
```suggestion
// InitCounterVec initializes and registers the counter metrics vector. It
// takes a list of BoundedLabel objects - if any label value or label name is
// nil, then this method will panic. Based on the combinations returned by
// generateCombinations, it will set each counter value to 0.
// If a nil counter is returned during registration, the method will panic.
```